File: src/backend/shared/middleware/auth.middleware.ts
// @ts-nocheck
import { Request, Response, NextFunction } from 'express';
import { AuthService, AuthError, JwtPayload } from '../services/AuthService';

/**
 * Expressstyle middleware that validates a Bearer JWT.
 * On success, attaches the decoded payload to `req.user`.
 * On failure, responds with 401 using the standardized error schema.
 *
 * Usage (in an Azure Functions HTTP trigger wrapper or any Express app):
 *   app.use(authMiddleware);
 */
export const authMiddleware = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
  try {
    const authHeader = req.headers['authorization'] ?? req.headers['Authorization'];
    if (!authHeader || !authHeader.toString().startsWith('Bearer ')) {
      res.status(401).json({ error: 'unauthorized', detail: 'Missing Authorization header', code: 401 });
      return;
    }

    const token = authHeader.toString().substring('Bearer '.length).trim();
    const authService = new AuthService();
    const payload: JwtPayload = await authService.validateJwt(token);

    // Attach to request for downstream handlers
    (req as any).user = payload;
    next();
  } catch (err: any) {
    if (err instanceof AuthError) {
      res.status(err.code).json({ error: 'auth_error', detail: err.message, code: err.code });
    } else {
      console.error('Unexpected error in auth middleware', err);
      res.status(500).json({ error: 'server_error', detail: 'Unexpected error', code: 500 });
    }
  }
};

File: backend/shared/middleware/auth.middleware.py
"""HTTP middleware for Azure Functions to enforce JWT authentication."""

import json
import logging
from typing import Callable, Any

logger = logging.getLogger(__name__)

class AuthMiddleware:
    """Callable middleware that validates JWT and injects claims into the request."""

    def __init__(self, auth_service):
        """
        Args:
            auth_service: Instance of AuthService providing validate_jwt().
        """
        self.auth_service = auth_service

    def __call__(self, func: Callable) -> Callable:
        """Wrap an Azure Function entry point."""
        def wrapper(req: Any, *args, **kwargs):
            auth_header = req.headers.get("Authorization", "")
            if not auth_header.startswith("Bearer "):
                return func.HttpResponse(
                    json.dumps({"error": "Missing or malformed Authorization header"}),
                    status_code=401,
                    mimetype="application/json",
                )
            token = auth_header.split(" ", 1)[1]
            try:
                claims = self.auth_service.validate_jwt(token)
                # Attach claims for downstream handlers
                if not hasattr(req, "route_params"):
                    req.route_params = {}
                req.route_params["user"] = claims
            except Exception as exc:
                logger.exception("JWT validation failed")
                return func.HttpResponse(
                    json.dumps({"error": "Invalid token", "details": str(exc)}),
                    status_code=401,
                    mimetype="application/json",
                )
            return func(req, *args, **kwargs)
        return wrapper
